File: app/src/main/java/com/lucasurbas/counter/ui/explore/UiExploreState.java
package com.lucasurbas.counter.ui.explore;

import android.support.annotation.Nullable;

import com.google.auto.value.AutoValue;

import java.util.ArrayList;
import java.util.List;

import com.lucasurbas.counter.ui.explore.model.UiHeaderItem;
import com.lucasurbas.counter.ui.explore.model.UiListItem;
import com.lucasurbas.counter.ui.explore.model.UiPostItem;

@AutoValue
public abstract class UiExploreState {

    abstract boolean getIsLoading();

    abstract List<UiListItem> getItemList();

    @Nullable
    abstract Throwable getError();

    private static UiExploreState.Builder builder() {
        return new AutoValue_UiExploreState.Builder()
                .isLoading(false)
                .itemList(new ArrayList<>());
    }

    abstract Builder toBuilder();

    static UiExploreState initialState() {
        return builder().build();
    }

    @AutoValue.Builder
    abstract static class Builder {

        abstract Builder isLoading(boolean value);

        abstract Builder itemList(List<UiListItem> value);

        abstract Builder error(Throwable value);

        abstract UiExploreState build();
    }

    public interface Part {

        UiExploreState computeNewState(UiExploreState previousState);

        class Loading implements UiExploreState.Part {

            @Override
            public UiExploreState computeNewState(UiExploreState previousState) {
                return previousState.toBuilder()
                        .isLoading(true)
                        .error(null)
                        .build();
            }
        }

        class Error implements UiExploreState.Part {

            private Throwable error;

            public Error(Throwable error) {
                this.error = error;
            }

            @Override
            public UiExploreState computeNewState(UiExploreState previousState) {
                return previousState.toBuilder()
                        .isLoading(false)
                        .error(error)
                        .build();
            }
        }

        class PostsLoaded implements UiExploreState.Part {

            @Override
            public UiExploreState computeNewState(UiExploreState previousState) {
                return previousState.toBuilder()
                        .isLoading(false)
                        .build();
            }
        }

        class Posts implements UiExploreState.Part {

            private List<UiPostItem> uiPostItemList;

            public Posts(List<UiPostItem> uiPostItemList) {
                this.uiPostItemList = uiPostItemList;
            }

            @Override
            public UiExploreState computeNewState(UiExploreState previousState) {
                return previousState.toBuilder()
                        .itemList(calculateListData(uiPostItemList))
                        .build();
            }

            private List<UiListItem> calculateListData(List<UiPostItem> uiPostItemList) {
                List<UiListItem> data = new ArrayList<>();
                data.add(UiHeaderItem.newInstance());
                data.addAll(uiPostItemList);
                return data;
            }
        }
    }
}

File: app/src/main/java/com/lucasurbas/counter/ui/explore/UiExploreState.kt
package com.lucasurbas.counter.ui.explore

import com.lucasurbas.counter.ui.explore.model.UiCounterItem
import com.lucasurbas.counter.ui.explore.model.UiHeaderItem
import com.lucasurbas.counter.ui.explore.model.UiListItem

data class UiExploreState(
        val itemList: List<UiListItem> = ArrayList(),
        val error: Throwable? = null
) {
    interface Part {

        fun computeNewState(previousState: UiExploreState): UiExploreState

        class Error(private val error: Throwable) : UiExploreState.Part {

            override fun computeNewState(previousState: UiExploreState): UiExploreState {
                return previousState.copy(error = error)
            }
        }

        class CounterList(private val uiCounterItemList: List<UiCounterItem>) : UiExploreState.Part {

            override fun computeNewState(previousState: UiExploreState): UiExploreState {
                return previousState.copy(error = null, itemList = calculateListData(uiCounterItemList))
            }

            private fun calculateListData(uiCounterItemList: List<UiCounterItem>): List<UiListItem> {
                val data = ArrayList<UiListItem>()
                data.add(UiHeaderItem())
                data.addAll(uiCounterItemList)
                return data
            }
        }
    }
}
